browser/brave_wallet/solana_provider_browsertest.cc

```diff
@@ -0,0 +1,179 @@
+          ASSERT_TRUE(success);
+          run_loop.Quit();
+        }));
+    run_loop.Run();
+  }
+
+  void AddAccount() {
+    base::RunLoop run_loop;
+    keyring_service_->AddAccount(
+        "Account 1", mojom::CoinType::SOL,
+        base::BindLambdaForTesting([&run_loop](bool success) {
+          ASSERT_TRUE(success);
+          run_loop.Quit();
+        }));
+    run_loop.Run();
+  }
+
+  void SetSelectedAccount(const std::string& address) {
+    base::RunLoop run_loop;
+    keyring_service_->SetSelectedAccount(
+        address, mojom::CoinType::SOL,
+        base::BindLambdaForTesting([&run_loop](bool success) {
+          EXPECT_TRUE(success);
+          run_loop.Quit();
+        }));
+    run_loop.Run();
+  }
+
+  void UserGrantPermission(bool granted) {
+    if (granted)
+      permissions::BraveWalletPermissionContext::AcceptOrCancel(
+          std::vector<std::string>{first_account}, web_contents());
+    else
+      permissions::BraveWalletPermissionContext::Cancel(web_contents());
+    ASSERT_EQ(EvalJs(web_contents(), "getConnectedAccount()",
+                     content::EXECUTE_SCRIPT_USE_MANUAL_REPLY)
+                  .ExtractString(),
+              granted ? first_account : "");
+  }
+
+  void CallSolanaConnect() {
+    ASSERT_TRUE(ExecJs(web_contents(), "solanaConnect()"));
+    base::RunLoop().RunUntilIdle();
+    ASSERT_TRUE(
+        brave_wallet::BraveWalletTabHelper::FromWebContents(web_contents())
+            ->IsShowingBubble());
+  }
+
+  void CallSolanaDisConnect() {
+    ASSERT_TRUE(EvalJs(web_contents(), "solanaDisconnect()",
+                       content::EXECUTE_SCRIPT_USE_MANUAL_REPLY)
+                    .ExtractBool());
+  }
+
+  bool IsSolanaConnected() {
+    return EvalJs(web_contents(), "isSolanaConnected()",
+                  content::EXECUTE_SCRIPT_USE_MANUAL_REPLY)
+        .ExtractBool();
+  }
+
+ private:
+  base::test::ScopedFeatureList scoped_feature_list_;
+  net::test_server::EmbeddedTestServer https_server_;
+  raw_ptr<KeyringService> keyring_service_ = nullptr;
+};
+
+IN_PROC_BROWSER_TEST_F(SolanaProviderTest, ConnectedStatusAndPermission) {
+  RestoreWallet();
+  AddAccount();
+  SetSelectedAccount(first_account);
+  GURL url = https_server()->GetURL("a.com", "/solana_provider.html");
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), url));
+  EXPECT_TRUE(WaitForLoadStop(web_contents()));
+
+  ASSERT_FALSE(IsSolanaConnected());
+  CallSolanaConnect();
+  UserGrantPermission(true);
+  EXPECT_TRUE(IsSolanaConnected());
+
+  // remove solana permission doesn't affect connected status
+  host_content_settings_map()->ClearSettingsForOneType(
+      ContentSettingsType::BRAVE_SOLANA);
+  EXPECT_TRUE(IsSolanaConnected());
+
+  // connect again and reject it doesn't affect connected status either
+  CallSolanaConnect();
+  UserGrantPermission(false);
+  EXPECT_TRUE(IsSolanaConnected());
+
+  // only solana.disconnect and set connected status to false
```

Comment:
nit: `Only disconnect will set connected status to false.`?

Reply:
fixed and rebased d5242b1743014b073130dbd1c1ca850aae4a0c90